### PR TITLE
Create an example for structured data classification with Neural Decision Forests

### PR DESCRIPTION
**Pull request by ksalama**
This example provides an implementation of the [Deep Neural Decision Forest](https://www.cv-foundation.org/openaccess/content_iccv_2015/papers/Kontschieder_Deep_Neural_Decision_ICCV_2015_paper.pdf)
model for structured data classification. This model introduced a stochastic and differentiable decision tree model to unify
classification trees with the deep representation learning, in a joint training routine.

### CODE REVIEW

**Comment by ksalama** [comment]
@fchollet @8bitmp3 

Hello - An other example for structure data classification with Neural Decision trees! 
Sorry for keeping you busy recently :-)

Cheers

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:104]: Add a comment above each global variable you define

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:136]: Why the leading space?

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:149]: target (typo)

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:169]: You could just do `lambda x, y: (x, target_label_lookup(y))` instead of having a separate function

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:171]: Consider caching / prefetching

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:353]: Line break above

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:279]: Add a paragraph describing the architecture

File: examples/structured_data/deep_neural_decision_forests.py
Comment (fchollet) [inline examples/structured_data/deep_neural_decision_forests.py:382]: Avoid quoting overly specific numbers since they differ with each run

**Review by fchollet — COMMENTED** [review]
This is another very interesting example. Thanks for the great contribution! 👍

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:12]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:75]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:113]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:104]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:169]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:171]: Done. I added `.cache()`. Note that the `tf.data.experimental.make_csv_dataset()` method performs prefetch automatically.

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:353]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:382]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:2]: Suggestion:

- Use less vague titles/definitions to help users understand what your example is about: "classification" -> "structured data classification"


```suggestion
Title: Structured data classification with Neural Decision Forests
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:6]: Suggestion:

- Use words from the author's summary/conclusion from the paper to help write more useful descriptions (descriptions shouldn't be just longer versions of titles IMO in this case). Help the users figure out why they should spend their time learning about this technique by scanning the first few paragraphs. The authors have already done the work.

```suggestion
Description: How to model and train stochastic, differentiable decision trees as classifiers for end-to-end learning in deep convolutional networks.
```

**Review by 8bitmp3 — COMMENTED** [review]
A few suggestions, based on what I read in the original dNDF paper:

```
Description: How to model and train stochastic, differentiable decision
trees as classifiers for end-to-end learning in deep convolutional networks.
```

```
Title: Image classification with Deep Neural Decision Forests
```

Hope this helps!

**Comment by 8bitmp3** [comment]
🎄 Happy Holidays @fchollet @ksalama ! 🔔

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:76]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:270]: Done

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:2]: In this example, we apply the model for structured data, rather than images.

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:6]: Thanks for the suggestion. I adapted your description but changed `deep convolutional networks` to `deep learning neural networks`, as we apply the model to structured data, rather images, thus we don't use convolutional networks.

**Comment by ksalama** [comment]
@fchollet and @8bitmp3 Thank you so much for your comments and suggestions.
I have made the requested modifications, including changing the title and the description of the tutorial, and adding a paragraph describing the Neural Decision Tree model. Please let me know what you think.

Happy Holidays!!!

**Review by fchollet — COMMENTED** [review]
I did a round of copyedits. Other than that, everything looks good! This is a great example, very clear and concise.

@8bitmp3 if you have the time, I'd appreciate a second look as well.

**Comment by ksalama** [comment]
@fchollet Thanks for your edits. I have formatted the code file. Will add the .ipynb and .md files as soon as I get your approval.

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:16]: Suggestion: refactor - repetition "model... model"

```suggestion
This example provides an implementation of the
[Deep Neural Decision Forest](https://ieeexplore.ieee.org/document/7410529)
model introduced by P. Kontschieder et al. for structured data classification.
It demonstrates how to build a stochastic and differentiable decision tree model,
train it end-to-end, and unify decision trees with deep representation learning.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:25]: Suggestion: thinking of a global audience

```suggestion
to predict whether a person is likely to be making over USD 50,000 a year.
```

("k" or "K" may not translate into 1,000s in all cultures, I think)

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:27]: Suggestion: just adding just a little bit more information about the dataset.

```suggestion
The dataset includes 48,842 instances with 14 input features (such as age, work class, education, occupation, and so on): 5 numerical features
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:79]: Suggestion: rephrase a little to avoid repetition ("remove...remove")

```suggestion
Remove the first record (because it is not a valid data example) and a trailing
'dot' in the class labels.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [reply examples/structured_data/deep_neural_decision_forests.py:137]: @ksalama Has this been clarified? Just checking

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:144]: I think this is "training and validation" (separately)

```suggestion
for training and validation. We also preprocess the input by mapping the target label
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:203]: Comma

```suggestion
            # Since we are not using a mask token, nor expecting any out of vocabulary
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:204]: Double space

```suggestion
            # (oov) token, we set mask_token to None and num_oov_indices to 0.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:251]: Nit (step 3 is not "second"):

```suggestion
1. The model expects input `features` as a single vector encoding all the features of an instance
in the batch. This vector can be generated from a Convolution Neural Network (CNN) applied to images
or dense transformations applied to structured data features. 
2. The model first applies a `used_features_mask` to randomly select a subset of input features to use.
3. Then, the model computes the probabilities (`mu`) for the input instances to reach the tree leaves
by iteratively performing a *stochastic* routing throughout the tree levels.
4. Finally, the probabilities of reaching the leaves are combined by the class probabilities at the
leaves to produce the final `outputs`.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:270]: 🇺🇸 English is international:

```suggestion
        # Initialize the weights of the classes in leaves.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:390]:
```suggestion
In this experiment, we train a single neural decision tree model
where we use all input features.
```

File: examples/structured_data/deep_neural_decision_forests.py
Comment (8bitmp3) [inline examples/structured_data/deep_neural_decision_forests.py:420]:
```suggestion
In this experiment, we train a neural decision forest with `num_trees` trees
where each tree uses randomly selected 50% of the input features. You can control the number
```

**Review by 8bitmp3 — CHANGES_REQUESTED** [review]
Thanks, done @fchollet  

PTAL @ksalama Hope this helps

File: examples/structured_data/deep_neural_decision_forests.py
Comment (ksalama) [reply examples/structured_data/deep_neural_decision_forests.py:137]: This is how the data is in the original dataset.

**Comment by ksalama** [comment]
@8bitmp3 thanks for suggestions. I have incorporated them in the code file.
I have now generated the .ipynb and .md files.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

**Comment by fchollet** [comment]
We get the warning: "UserWarning: Input dict contained keys ['fnlwgt'] which did not match any model input. They will be ignored by the model". We should just pop this feature so it doesn't get included in the input dict.

**Pull request merged**
